### DISCUSSION TITLE
Warning ao declarar o DbSet<TodoModel> de valor nulo

### DISCUSSION THREAD

**Discussion by maransi** [balta-io/forum#11] [Tira Dúvidas] [answered]
Boa noite Balta.

No curso "Fundamentos do ASP.NET 6", estou enfrentando um problema ao declarar o DbSet<TodoModel> no arquivo "AppDbContext". 

Ao fazer a declaração:

            public DbSet<TodoModel> Todos { get; set; }

Estou recebendo um aviso de "warning": 

(A propriedade não anulável 'Todos' precisa conter um valor não nulo ao sair do construtor. Considere declarar o propriedade como anulável).

Pesquisando consegui eliminar o aviso colocando o operador (?): 

           public DbSet<TodoModel>? Todos { get; set; }

Resolveu parcialmente, porém na classe "HomeController" todas as vezes que tento me referenciar ao DbSet Todos através de injeção de dependência a mesma mensagem de valor nulo acaba saindo:

        [HttpGet]
        public List<TodoModel> Get([FromServices] AppDbContext context )
        {
            return context.Todos.ToList();  
            // Msg warning: Possível argumento de referência nula para o parâmetro 'source' em 'List<TodoModel> 
            //                         Enumerable.ToList<TodoModel>(IEnumerable<TodoModel> source)'. [Todo]csharp(CS8604)
        }

Pesquisando descobri que é possível colocar um operador "nulável" (!) na chamado do contexto e a msg desaparece:

            return context.Todos!.ToList();  

Esse é o meu primeiro projeto com o .NET 6, com as versões anteriores não encontrava esse tipo de problema. É assim mesmo que devemos fazer de agora em diante? não tem uma outra maneira de suprimir essa msg? vc já enfrentou esse erro? Nos vídeos do seu curso vc utiliza o .NET 6 e não vi esse erro ocorrer.

Obrigado.

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @maransi, como vai?

Acredito que o curso está na versão 5 do .NET e você deve estar na versão 6.
Dito isto, no .NET 6 os projetos vem com um parâmetro chamado `<nullable>` ativado no `.csproj`:
```xml
<nullable>true</nullable>
```

Isto causa estes **warnings** que está tendo. O compilador está te dizendo que você não declarou um tipo como nulo e nem disse como ele deve ser inicializado, então pode ocorrer algum erro (Null Reference).

Se você sabe que este objeto nunca vai ser nulo, você pode usar isto (Que eu recomendaria neste caso):
```csharp
public DbSet<MinhaClasse> MinhasClasses { get; set; } = null!;
```

Outras opções:
```csharp
public DbSet<MinhaClasse> MinhasClasses { get; set; } = new();
public DbSet<MinhaClasse>? MinhasClasses { get; set; }
```

**Comment by maransi** [comment]
Bom dia Balta.

"A melhor coisa em ficar até altas horas da noite quebrando a cabeça com um problema , é chegar no dia seguinte e ter a solução, rsrsrs"

Obrigado pelas saídas que vc postou. No meu caso a que se encaixou melhor foi:

      <nullable>true</nullable>

Só um detalhe para completar, ao testar a maneira correta que encontrei foi:

    <nullable>disable</nullable>

Completando, as possibilidades abaixo funcionaram num primeiro momento, mas depois ao usar o objeto de contexto nos "controllers" por exemplo acaba ocorrendo o mesmo problema de objetos nulos:

    public DbSet<MinhaClasse> MinhasClasses { get; set; } = null!;
    public DbSet<MinhaClasse>? MinhasClasses { get; set; }

Já a opção abaixo, infelizmente não consegui fazer funcionar:

    public DbSet<MinhaClasse> MinhasClasses { get; set; } = new();

Valeu mesmo.

**Reply by andrebaltieri** [reply]
Show demais!!!!
